File: src/main/java/org/voegtle/weatherstation/server/logic/WeatherDataAggregator.java
package org.voegtle.weatherstation.server.logic;

import org.voegtle.weatherstation.server.persistence.*;
import org.voegtle.weatherstation.server.persistence.entities.AggregatedWeatherDataSet;
import org.voegtle.weatherstation.server.persistence.entities.SmoothedWeatherDataSet;
import org.voegtle.weatherstation.server.persistence.entities.WeatherDataSet;
import org.voegtle.weatherstation.server.util.DateUtil;

import java.util.Date;
import java.util.List;
import java.util.logging.Logger;

public class WeatherDataAggregator {
  private static final Logger log = Logger.getLogger(WeatherDataAggregator.class.getName());

  private final PersistenceManager pm;
  private final DateUtil dateUtil;

  public WeatherDataAggregator(PersistenceManager pm, DateUtil dateUtil) {
    this.pm = pm;
    this.dateUtil = dateUtil;
  }

  public void aggregateWeatherData() {
    Date dateOfLastAggregation = fetchDateOfLastAggregation();
    Date dateOfLastWeatherDataSet = fetchLastDateWithCompleteWeatherDataSets();

    while (dateUtil.isClearlyBefore(dateOfLastAggregation, dateOfLastWeatherDataSet)) {
      AggregatedWeatherDataSet aggregatedDay = createNewDay(dateOfLastAggregation);
      log.warning("aggregate " + aggregatedDay.getDate());
      List<SmoothedWeatherDataSet> weatherDataSets = pm.fetchSmoothedWeatherDataInRange(dateUtil.fromLocalToGMT(aggregatedDay.getDate()),
          dateUtil.fromLocalToGMT(dateUtil.nextDay(aggregatedDay.getDate())));

      aggregate(aggregatedDay, weatherDataSets);

      pm.makePersitant(aggregatedDay);
      dateOfLastAggregation = aggregatedDay.getDate();
    }
  }

  private void aggregate(AggregatedWeatherDataSet aggregation, List<SmoothedWeatherDataSet> weatherDataSets) {
    if (weatherDataSets.size() > 0) {
      Integer rainCountStart = null;
      Integer rainCountLast = null;

      Double kwhStart = null;
      Double kwhLast = null;
      for (SmoothedWeatherDataSet wds : weatherDataSets) {
        if (wds.isValid()) {
          log.warning("Smoothed DS: " + wds.getTimestamp());
          aggregation.addOutsideTemperature(wds.getOutsideTemperature(), wds.getTimestamp());
          aggregation.addOutsideHumidity(wds.getOutsideHumidity());
          aggregation.addInsideTemperature(wds.getInsideTemperature());
          aggregation.addInsideHumidity(wds.getInsideHumidity());
          if (wds.getRainCounter() != null) {
            if (rainCountStart == null) {
              rainCountStart = wds.getRainCounter();
            }
            rainCountLast = wds.getRainCounter();
          }
          if (kwhStart == null) {
            kwhStart = wds.getKwh();
          }
          kwhLast = wds.getKwh();
        }
      }

      if (rainCountStart != null) {
        aggregation.setRainCounter(Math.max(rainCountLast - rainCountStart, 0));
        aggregation.setRainDays(rainCountLast > rainCountStart ? 1 : 0);
      }

      if (kwhStart != null && kwhLast != null) {
        aggregation.setKwh(Math.max(kwhLast - kwhStart, 0));
      }

      aggregation.normalize();
    }
    aggregation.setFinished(true);
  }

  private Date fetchDateOfLastAggregation() {
    AggregatedWeatherDataSet lastAggregatedDay = pm.fetchYoungestAggregatedDataSet(PeriodEnum.DAY);
    return lastAggregatedDay == null ? dateUtil.getDate(2016, 5, 11) : lastAggregatedDay.getDate();
  }

  private Date fetchLastDateWithCompleteWeatherDataSets() {
    WeatherDataSet youngest = pm.fetchYoungestDataSet();
    Date timestamp = dateUtil.daysEarlier(youngest.getTimestamp(), 1);
    timestamp = dateUtil.fromGMTtoLocal(timestamp);
    return timestamp;
  }

  private AggregatedWeatherDataSet createNewDay(Date lastDay) {
    return new AggregatedWeatherDataSet(dateUtil.incrementDay(lastDay), PeriodEnum.DAY);
  }
}

File: src/main/java/org/voegtle/weatherstation/server/logic/WeatherDataAggregator.kt
package org.voegtle.weatherstation.server.logic

import org.voegtle.weatherstation.server.persistence.PersistenceManager
import org.voegtle.weatherstation.server.persistence.entities.AggregatedWeatherDataSet
import org.voegtle.weatherstation.server.persistence.entities.SmoothedWeatherDataSet
import org.voegtle.weatherstation.server.util.DateUtil
import java.util.*
import java.util.logging.Logger

class WeatherDataAggregator(private val pm: PersistenceManager, private val dateUtil: DateUtil) {
  private val log = Logger.getLogger(WeatherDataAggregator::class.java.name)

  fun aggregateWeatherData() {
    var dateOfLastAggregation = fetchDateOfLastAggregation()
    val dateOfLastWeatherDataSet = fetchLastDateWithCompleteWeatherDataSets()

    while (dateUtil.isClearlyBefore(dateOfLastAggregation, dateOfLastWeatherDataSet)) {
      val aggregatedDay = createNewDay(dateOfLastAggregation)
      log.info("aggregate " + aggregatedDay.date)
      val weatherDataSets = pm.fetchSmoothedWeatherDataInRange(dateUtil.fromLocalToGMT(aggregatedDay.date),
                                                               dateUtil.fromLocalToGMT(dateUtil.nextDay(aggregatedDay.date)))

      aggregate(aggregatedDay, weatherDataSets)

      pm.makePersistent(aggregatedDay)
      dateOfLastAggregation = aggregatedDay.date
    }
  }

  private fun aggregate(aggregation: AggregatedWeatherDataSet, weatherDataSets: List<SmoothedWeatherDataSet>) {
    if (weatherDataSets.isNotEmpty()) {
      for (wds in weatherDataSets) {
        if (wds.isValid) {
          aggregation.addOutsideTemperature(wds.outsideTemperature, wds.timestamp)
          aggregation.addOutsideHumidity(wds.outsideHumidity)
          aggregation.addInsideTemperature(wds.insideTemperature)
          aggregation.addInsideHumidity(wds.insideHumidity)
          aggregation.addWindspeed(wds.windspeed)
          aggregation.updateWindspeedMax(wds.windspeedMax)
          aggregation.addSolarRadiation(wds.solarRadiation)
          aggregation.updateSolarRadiationMax(wds)
          aggregation.totalPowerProduction = wds.totalPowerProduction
          aggregation.updatePowerProductionMax(wds.powerProductionMax, wds.timestamp)
          aggregation.dailyRain = wds.dailyRain
        }
      }
      aggregation.totalPowerProduction = calculateDailyPowerProduction(weatherDataSets)
      aggregation.normalize()
    }
  }

  private fun calculateDailyPowerProduction(weatherDataSets: List<SmoothedWeatherDataSet>): Float? {
    var totalPowerProductionInitial = weatherDataSets.first().totalPowerProduction
    var totalPowerProductionFinal = weatherDataSets.last().totalPowerProduction
    if (totalPowerProductionInitial != null && totalPowerProductionFinal != null) {
      return (totalPowerProductionFinal - totalPowerProductionInitial).coerceAtLeast(0.0f)
    }
    return null
  }

  private fun fetchDateOfLastAggregation(): Date {
    val lastAggregatedDay = pm.fetchYoungestAggregatedDataSet()
    return lastAggregatedDay?.date ?: dateUtil.getDate(2021, 1, 1)
  }

  private fun fetchLastDateWithCompleteWeatherDataSets(): Date {
    val youngest = pm.fetchYoungestDataSet()
    var timestamp = dateUtil.daysEarlier(youngest.time, 1)
    timestamp = dateUtil.fromGMTtoLocal(timestamp)
    return timestamp
  }

  private fun createNewDay(lastDay: Date): AggregatedWeatherDataSet {
    return AggregatedWeatherDataSet(date = dateUtil.incrementDay(lastDay))
  }

}
